### PR TITLE
`MatrixMorphism_abstract`: Move endomorphism invariants to category

### PR DESCRIPTION
**Pull request by mkoeppe**
We move the methods `charpoly`, `det`, `fcp`, `trace` from `MatrixMorphism_abstract` to the `Homsets.Endset.ElementMethods` of the category `FiniteDimensionalModulesWithBasis`. Likewise, we move the method `minpoly` there from `FreeModuleMorphism`.

This makes them also available to endomorphisms of `CombinatorialFreeModule`.

We also deprecate the functions `is_MatrixMorphism`, `is_FreeModuleMorphism`, `is_VectorSpaceMorphism` (as part of #32414); they were almost unused.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by tscrim** [comment]
There is an `Endsets` axiom for Homsets, so you can move those methods to that subcatregory.

**Comment by mkoeppe** [comment]
https://github.com/sagemath/sage/blob/develop/src/sage/categories/modules.py#L818 seems to caution that `Endset` "is not used yet", but I'll try if this works.

File: src/sage/modules/free_module_morphism.py [hunk 1]
```diff
@@ -49,18 +50,31 @@
...
+    from sage.misc.superseded import deprecation
+    deprecation(37731,
+                "is_FreeModuleMorphism is deprecated; "
+                "use isinstance(..., FreeModuleMorphism) or categories instead")
     return isinstance(x, FreeModuleMorphism)
 
 
 class FreeModuleMorphism(matrix_morphism.MatrixMorphism):
+
+    minimal_polynomial = minpoly = FiniteDimensionalModulesWithBasis.Homsets.Endset.ElementMethods.minimal_polynomial
```
Comment (tscrim) [inline src/sage/modules/free_module_morphism.py]: Why is this necessary? Shouldn't this come from the category?

Comment (mkoeppe) [reply src/sage/modules/free_module_morphism.py]: I've removed it

**Comment by mkoeppe** [comment]
Any further concerns, or can we merge it?

**Review by tscrim — APPROVED** [review]
Yes, I think this can be merged.

**Comment by mkoeppe** [comment]
Thanks!
